### ISSUE TITLE
`ERROR: CONDA_BUILD_SYSROOT or SDKROOT has to be set for cross-compiling`

### ISSUE THREAD

**Issue by hblumberg** [OceanParcels/parcels#1299]
Thank you so much for adding `osx-arm64` support in v2.4.0!

We tried creating a conda environment on an Apple M1: `conda create -n test-env -c conda-forge parcels`

`conda activate test-env` gives the following error:
```
ERROR: CONDA_BUILD_SYSROOT or SDKROOT has to be set for cross-compiling
activate_clang:69: read-only file system: /meson_cross_file.txt
activate_clang:70: read-only file system: /meson_cross_file.txt
activate_clang:71: read-only file system: /meson_cross_file.txt
activate_clang:72: read-only file system: /meson_cross_file.txt
activate_clang:73: read-only file system: /meson_cross_file.txt
```

We asked about this error [here](https://github.com/conda-forge/clang-compiler-activation-feedstock/issues/94) and they suggested that the `parcels` recipe might be wrong. They linked to https://github.com/conda-forge/parcels-feedstock/blob/main/recipe/meta.yaml#L38.

We thought maybe `- clang_osx-arm64` needed to be added to the recipe, but we noticed that there's already a [commit](https://github.com/conda-forge/parcels-feedstock/pull/85/commits/8b50d230d9e2a61eb911cc8fb0b5554c621cdd7c) that removed `- clang_osx-arm64` after adding it.

Do you have any further guidance on this issue? Thanks again!

**Comment by erikvansebille** [comment]
Thanks for reporting, @hblumberg. Not sure that I have more guidance, but perhaps @willirath has experience with `osx-arm64`?

**Comment by erikvansebille** [comment]
Now that we have a `no-arch` version on conda since https://github.com/conda-forge/parcels-feedstock/pull/98, my experience is that macOS works easier. So closing this issue (for now)

**Comment by vbalza** [comment]
I think there is a no-arch version now:
https://github.com/OceanParcels/parcels/issues/1299#issuecomment-1655662996

Perhaps we can revisit in an upcoming sprint.

**Comment by erikvansebille** [comment]
Thanks @JamiePringle, for providing this patch. But rather than adding another `environment*.yml` file, I wonder if we could not go down the route of https://github.com/conda-forge/parcels-feedstock/pull/98 and provide only one environment file for all architectures. 

A bit like how it's done in 
https://github.com/conda-forge/parcels-feedstock/blob/50ab3a48c31416a926185101cfbb339de49940d5/recipe/meta.yaml#L21-L48

Does @willirath, @ocefpaf or @ngam perhaps know how to do that for the `environment*.yml` files in https://github.com/OceanParcels/parcels that conda uses, too?

**Comment by ocefpaf** [comment]
These two lines,

https://github.com/conda-forge/parcels-feedstock/blob/50ab3a48c31416a926185101cfbb339de49940d5/recipe/meta.yaml#L36-L37

ensures that compilers will be available for Windows and "not windows." However, I'm not sure the `[unix]` pre-processor selector covers `osx-arm64`, probably not b/c that arch is still kind of new. I'll ask this in the conda forums.

With that said we can:

1. add another line with the proper compiler for `osx-arm64` (probably would require some work upstream to recognize that as OS-noarch)
2. document that users are responsible for installing the compiler
3. go back to the drawing board and make the recipe with multiple outputs so we can have a core parcels `noarch ` and an arch'ed parcels package that adds the compilers and depends on the core one.


IMO 3 is the best solution b/c it doesn't depend on any action from upstream.

**Comment by erikvansebille** [comment]
I've created #1416 to test if we can move to an environment file that doesn't include compilers. If this passes on the Github Action CI, then we should try to implement this?

**Comment by erikvansebille** [comment]
> I've created https://github.com/OceanParcels/parcels/pull/1416 to test if we can move to an environment file that doesn't include compilers. If this passes on the Github Action CI, then we should try to implement this?

OK, so removing the compilers from each of the three environment files doesn't break the CI (don't mind the red cross in https://github.com/OceanParcels/parcels/pull/1416; that's because of a strange bug in ReadTheDocs, see #1417). So that seems that we can even have an option 4 on @ocefpaf's list above; or am I missing something?

**Comment by ocefpaf** [comment]
Could your tests be picking up any compilers available in the CIs? If so, that won't help end users, unless you amend the docs and tell them to install compilers from the system or with conda.

**Comment by erikvansebille** [comment]
> Could your tests be picking up any compilers available in the CIs? If so, that won't help end users, unless you amend the docs and tell them to install compilers from the system or with conda.

Well, this is for the Developers version of the environments, so we can expect them to have a compiler available (and will explain that in the documentation at https://docs.oceanparcels.org/en/latest/installation.html#installation-for-developers

'Normal' users would use the condo package at https://github.com/conda-forge/parcels-feedstock, so would not deal with these environment files anyways

**Comment by erikvansebille** [comment]
Hmm, adding `[unix]` behind the `mpi4py` and `mpich` packages (because they don't exist for windows doesn't seem to work in https://github.com/OceanParcels/parcels/pull/1416/commits/079777c0480d479010382ec96e542f0b31b7821d. 

```python
  - mpi4py>=3.0.1 # [unix]
  - mpich>=3.2.1  # [unix]
 ```
This gives an error in [Windows CI that mpich does not exist](https://github.com/OceanParcels/parcels/actions/runs/5967433791/job/16189102347?pr=1416)

Anyone has an idea how to specify that certain packages should only be installed under unix architectures in a conda environment.yml file?

**Comment by ocefpaf** [comment]
Conda doesn't support that. It is a long standing issue. See https://github.com/conda/conda/issues/8089

You can do it with mamba/micromamba though. See https://github.com/mamba-org/mamba/blob/20b477f4595e7d932034945fd122a5a40cd25c9c/libmamba/tests/data/env_file/env_2.yaml#L6 for an example.

**Comment by erikvansebille** [comment]
OK, thanks for this information, @ocefpaf. 

And would you have any advice on whether it's worth it to change to mamba/micromamba for the developer installation then? I don't have much experience with it, but from what I read it seems a newer/better implementation of conda?

**Comment by ocefpaf** [comment]
I highly recommend micromamba. Now, with latest conda adopting the libmamba solver as its default solver, the solve times for all 3 are virtually the same and for most users those are equivalent However, micromamba is a smaller download, single binary (no Python base env to mess with), and faster overall for things beyond the solver.

**Comment by erikvansebille** [comment]
OK, so how could we then change the https://github.com/OceanParcels/parcels/blob/noacrch-environment-file/environment.yml file so that it works with mamba? 

And do we then also need to change the https://github.com/OceanParcels/parcels/blob/noacrch-environment-file/.github/workflows/integration-tests.yml and https://github.com/OceanParcels/parcels/blob/noacrch-environment-file/.github/workflows/unit-tests.yml files so that they use mamba and work with this new environment file?

Could you help us develop this in https://github.com/OceanParcels/parcels/pull/1416?

**Comment by ocefpaf** [comment]
> You will need to somehow tell the solver about different needs on different systems.

You can by using `sel(arch)`, for example:

```yaml
name: TEST
channels:
  - conda-forge
dependencies:
  - python>=3.8
  - sel(unix): mpich>=3.2.1  # will be installed only on not-windows
  - sel(win): colorama  # will be installed only on windows
```

Is that what you want @erikvansebille ?

**Comment by erikvansebille** [comment]
> You can by using `sel(arch)`
> 
> Is that what you want @erikvansebille ?

Thanks for this suggestion, @ocefpaf! It works for Github Action, seehttps://github.com/OceanParcels/parcels/pull/1416/commits/e8be1af07aeb3223cf1b0b7cfcb925d05a486700

However, the build for ReadtheDocs now fails, using the same environment.yml file. See https://readthedocs.org/projects/parcels/builds/21739117/
```
mamba env create --quiet --name 1416 --file environment.yml
Warning: you have pip-installed dependencies in your environment file, but you do not list pip itself as one of your conda dependencies.  Conda may not use the correct pip to install your packages, and they may end up in the wrong place.  Please add an explicit pip dependency.  I'm adding one for you, but still nagging you.
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
Unable to install package for sel(unix).

Please double check and ensure your dependencies file has
the correct spelling.  You might also try installing the
conda-env-sel(unix) package to see if provides the required
installer.
Retrieving notices: ...working... done
```

Any idea why the `mamba` implementation that readthedocs uses doesn't accept these `sel(unix)` decorators? Am I missing something?

**Comment by ocefpaf** [comment]
> Any idea why the `mamba` implementation that readthedocs uses doesn't accept these `sel(unix)` decorators? Am I missing something?

Do you know what version they are using? It may be a matter of asking them to update it.

**Comment by erikvansebille** [comment]
> Do you know what version they are using? It may be a matter of asking them to update it.

It's mambaforge-22.9.0-3, see also the screenshot below from https://readthedocs.org/projects/parcels/builds/21739117/
<img width="809" alt="Screenshot 2023-08-28 at 14 48 16" src="https://github.com/OceanParcels/parcels/assets/14315062/695b44fa-5123-478e-ae38-f37b4d5a3b1e">

Is that indeed an older version where `sel(unix)` is not implemented yet?

**Comment by ocefpaf** [comment]
> Is that indeed an older version where sel(unix) is not implemented yet?

Mambaforge 22.9.0 uses mamba mamba 1.1.0 but latest is 1.5.0. If I had to guess, yes, that one is too old. But I could not find any [mentions of it in the changelog](https://github.com/mamba-org/mamba/blob/main/CHANGELOG.md).

I'm not experienced with readthedocs to know for sure but can you control and/or install/update the underlying conda flavor in it?

Latest is 23.3.1

**Comment by erikvansebille** [comment]
This is now fixed in #1416
